src/mprispropertiesadaptor.cpp

```diff
@@ -0,0 +1,240 @@
+
+        if (pidx != -1) {
+            QMetaProperty mp = m_playerPrivate->q_ptr->metaObject()->property(pidx);
+            result = mp.read(m_playerPrivate->q_ptr);
+        } else {
+            replyInternalError();
+        }
+    }
+
+    return result;
+}
+
+QVariantMap MprisPropertiesAdaptor::GetAll(const QString &interface_name)
+{
+    QVariantMap result;
+    QMap<QString, QString> getMap;
+
+    if (interface_name == QLatin1String("org.mpris.MediaPlayer2.Player")) {
+        getMap = playerGetMap;
+    } else if (interface_name == QLatin1String("org.mpris.MediaPlayer2")) {
+        getMap = serviceGetMap;
+    } else {
+        replyPropertyNotFoundError(interface_name, "");
+    }
+
+    QMap<QString, QString>::const_iterator it = getMap.constBegin();
+    while (it != getMap.constEnd()) {
+        QVariant value = get(it.value());
+        result.insert(it.key(), value);
+        ++it;
+    }
+
+    return result;
+}
+
+// Format: {DBus property name, method to call in MprisPlayerPrivate}
+static const QMap<QString, QString> playerSetMap{
+    {"LoopStatus", "setLoopStatus(QString)"},
+    {"Rate", "setRate(double)"},
+    {"Shuffle", "setShuffle(bool)"},
+    {"Volume", "setVolume(double)"},
+};
+
+static const QMap<QString, QString> serviceSetMap{
+    {"Fullscreen", "setFullscreen(bool)"},
+};
+
+void MprisPropertiesAdaptor::Set(const QString &interface_name, const QString &property_name, const QDBusVariant &value)
+{
+    QMap<QString, QString> setMap;
+    QMap<QString, QString> getMap;
+
+    if (interface_name == QLatin1String("org.mpris.MediaPlayer2.Player")) {
+        setMap = playerSetMap;
+        getMap = playerGetMap;
+    } else if (interface_name == QLatin1String("org.mpris.MediaPlayer2")) {
+        setMap = serviceSetMap;
+        getMap = serviceGetMap;
+    }
+
+    const QString &setter = setMap.value(property_name);
+    if (!setter.isEmpty()) {
+        set(setter, value);
+    } else {
+        if (getMap.contains(property_name)) {
+            replyPropertyReadOnlyError(interface_name, property_name);
+        } else {
+            replyPropertyNotFoundError(interface_name, property_name);
+        }
+    }
+}
+
+void MprisPropertiesAdaptor::set(const QString &setter, const QDBusVariant &value)
+{
+    const QMetaObject *metaobject = m_playerPrivate->metaObject();
+    bool success = false;
+
+    int pidx = metaobject->indexOfMethod(
+                setter.toLocal8Bit().data());
+
+    if (pidx != -1) {
+        QMetaMethod method = metaobject->method(pidx);
+        Q_ASSERT(method.parameterCount() == 1);
+
+        // QMetaType maps to Type, as long as our methods stick to the standard types
+        if (method.parameterType(0) == static_cast<int>(value.variant().type())) {
+            success = method.invoke(m_playerPrivate, Qt::DirectConnection,
+                                    QGenericArgument(method.parameterNames()[0],
+                                    value.variant().data()));
```

Comment:
Should this be indented a little bit more as it's an argument for `QGenericArgument`?

Reply:
Yes it should! Thanks, I've fixed it.